File: leetcode/1-two-sum.py
# https://leetcode.com/problems/two-sum/ (top-interview-questions)

from typing import List

# Time: O(n), Space: O(n)
def twoSum(nums: List[int], target: int) -> List[int]:
    nums_index = {}

    for i, n in enumerate(nums):
        complement = target - n
        if complement in nums_index:
            return [nums_index[complement], i]
        nums_index[n] = i

File: leetcode/1-two-sum.js
// https://leetcode.com/problems/two-sum/ (top-interview-questions)

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
  const complement = new Map();
  for (let i = 0; i < nums.length; i++) {
    const diff = target - nums[i];
    if (complement.has(diff)) {
      return [complement.get(diff), i];
    }
    complement.set(nums[i], i);
  }
};